### PR TITLE
licensecheck: more Scan adjustments, add MIT variants

### PR DESCRIPTION
**Pull request by rsc**

### CODE REVIEW

**Comment by robpike** [comment]
A CL like "more adjustments" is not sufficiently explanatory.

**Comment by rsc** [comment]
I'm sure I'm just not using GitHub "correctly". The individual commits spell out the actual changes.
(See the Commits tab: https://github.com/google/licensecheck/pull/29/commits).

Reproducing the descriptions here:

```
commit 85e9b24f3c91c2205cfffde4ac5fd9b39fc23dae (rsc/master)
Author:     Russ Cox <rsc@golang.org>
AuthorDate: Sun Aug 2 09:30:43 2020 -0400
Commit:     Russ Cox <rsc@golang.org>
CommitDate: Mon Aug 3 17:03:41 2020 -0400

    licenses: add MIT variants: MIT-0, MITNFA, JSON [LRE]

commit 514a1c1b14b767831b459bcd696788ec3cfbc57b
Author:     Russ Cox <rsc@golang.org>
AuthorDate: Sun Aug 2 09:55:10 2020 -0400
Commit:     Russ Cox <rsc@golang.org>
CommitDate: Mon Aug 3 16:20:13 2020 -0400

    internal/match: refactor re.compile1 into reCompile struct
    
    Using a reCompile struct is clearer and avoids the possibility
    of forgetting to save the result of an append to the prog.
    It also makes it easier to add new compilation state.

commit e258efa7815b4258f108a9d897001eacb5dcdc9f
Author:     Russ Cox <rsc@golang.org>
AuthorDate: Fri Jul 31 16:28:24 2020 -0400
Commit:     Russ Cox <rsc@golang.org>
CommitDate: Mon Aug 3 16:20:13 2020 -0400

    internal/match: reject licenses with short or wildcard leading phrases
    
    1-word leading phrases or 2-word phrases containing wildcards
    were silently dropped before the DFAs were all merged together.
    Now they bloat the DFA but are still unreachable because of the
    phrase search optimization during the main scan.
    Reject them instead.

commit 7e34d343a8bb1a2cf354bc61578771533c0771f7
Author:     Russ Cox <rsc@golang.org>
AuthorDate: Mon Aug 3 15:15:29 2020 -0400
Commit:     Russ Cox <rsc@golang.org>
CommitDate: Mon Aug 3 16:20:13 2020 -0400

    internal/match: recognize notice(s) as notices
    
    This construct happens in many licenses.
    Recognize it as one word and let the spell-checker
    deal with the singular/plural distinction.

commit d0c4dae03dff8aca1812f07b7ef6d98b8a907536
Author:     Russ Cox <rsc@golang.org>
AuthorDate: Sun Aug 2 09:49:39 2020 -0400
Commit:     Russ Cox <rsc@golang.org>
CommitDate: Mon Aug 3 16:03:33 2020 -0400

    licensecheck: initialize scanner lazily
    
    The DFA compilation can take a significant amount of time.
    Initializing it lazily makes that work appear in testing profiles
    (generated by -test.cpuprofile, which doesn't cover init-time work).

commit 30c85b6b5da6824bbd93359d271c5c0ab9fc5c13
Author:     Russ Cox <rsc@golang.org>
AuthorDate: Sun Aug 2 09:35:51 2020 -0400
Commit:     Russ Cox <rsc@golang.org>
CommitDate: Sun Aug 2 09:36:25 2020 -0400

    internal/match: change TraceDFA from bool to int
    
    This gives more control over how chatty we want the traces to be.
```

File: internal/match/regexp.go [hunk 1]
```diff
@@ -135,13 +134,31 @@ func NewMultiLRE(list []*LRE) (_ *MultiLRE, err error) {
 			return nil, fmt.Errorf("%s: no leading phrases", sub.File())
 		}
 		for _, p := range phrases {
+			if p[0] == BadWord {
+				return nil, fmt.Errorf("%s: license can be empty text", sub.File())
```
Comment (robpike) [inline internal/match/regexp.go:138]: the phrasing here threw me. you're speaking of what could happen if this pattern is accepted. it's really not important, but if you're up for it try something like this:

pattern accepts illegal license: cannot be empty
pattern accepts illegal license: cannot be a single wildcard
etc.

Comment (rsc) [reply internal/match/regexp.go]: Done.

**Pull request merged**